CONFIGURING.md
# Configuring

After loading the server with the plugin, the `config-hub.json` file should be created in directory `config`. <br>
The plugin configuration file uses the JSON format, see [json.org](https://www.json.org/json-en.html) site

### Available Fields

| Field           | Type                                      | Description                                                                                             |
| --------------- | ----------------------------------------- | ------------------------------------------------------------------------------------------------------- |
| offline-pattern | string                                    | pattern used when servers are offline                                                                   |
| online-pattern  | string                                    | placeholder based ([placeholders](#online-pattern-placeholders)) pattern, used when servers are online  |
| log-connects    | boolean                                   | log player connections to hub's servers in `[player uuid] player name --> server ip:server port` format |
| servers         | array of [server](#server-object) objects | hub's servers                                                                                           |
| effects         | array of [effect](#effect-object) objects | static effects that appear with a specified period                                                      |
| event-effects   | [event effects](#event-effect-object)     | effects called on a specific event                                                                      |

#### Online Pattern Placeholders

| Name          | Description                                             |
| ------------- | ------------------------------------------------------- |
| %address%     | server ip                                               |
| %mapname%     | current server map name                                 |
| %description% | server description                                      |
| %wave%        | current server map wave                                 |
| %players%     | server player count                                     |
| %playerLimit% | server player limit                                     |
| %version%     | server game version                                     |
| %versionType% | server game version type                                |
| %mode%        | server mode i.e. survival, sandbox, attack, pvp, editor |
| %modeName%    | server custom mode name, if absent, %mode% is used      |
| %ping%        | server ping                                             |
| %port%        | server port                                             |

### Server Object

###### Server Object structure

| Field      | Type    | Description                                                        |
| ---------- | ------- | ------------------------------------------------------------------ |
| ip         | string  | server ip                                                          |
| port       | integer | server port                                                        |
| size       | integer | the teleport block size                                            |
| teleport-x | integer | latter x coordinate of the teleport border. minimap coordinate     |
| teleport-y | integer | latter y coordinate of the teleport border. minimap coordinate     |
| title      | string  | server display name                                                |
| title-x    | float   | latter x coordinate of the title signboard. minimap coordinate * 8 |
| title-y    | float   | latter y coordinate of the title signboard. minimap coordinate * 8 |
| label-x    | float   | latter x coordinate of the status label. minimap coordinate * 8    |
| label-y    | float   | latter y coordinate of the status label. minimap coordinate * 8    |

###### Optimal Setup Formula

```
title-x = (teleport-x * 8) + 12
title-y = (teleport-y * 8) + 36

label-x = (teleport-x * 8) + 12
label-y = (teleport-y * 8) - 16
```

###### Example Server Object

```json
{
  "ip": "localhost",
  "port": 6567,
  "size": 7,
  "teleport-x": 22,
  "teleport-y": 37,
  "title": "title",
  "title-x": 200,
  "title-y": 320,
  "label-x": 200,
  "label-y": 288
}
```

### Effect Object

###### Effect Object structure

| Field         | Type    | Description                                                                                                                                      |
| ------------- | ------- | ------------------------------------------------------------------------------------------------------------------------------------------------ |
| x             | float   | effect x coordinate. minimap coordinate * 8                                                                                                      |
| y             | float   | effect y coordinate. minimap coordinate * 8                                                                                                      |
| rotation      | float   | effect rotation                                                                                                                                  |
| period-millis | integer | interval in milliseconds after which the effect will spawn                                                                                       |
| color         | string  | effect color in hex format                                                                                                                       |
| effect        | string  | effect name. the name of one of the fields of the [Fx](https://github.com/Anuken/Mindustry/blob/master/core/src/mindustry/content/Fx.java) class |

###### Example Effect Object

```json
{
  "x": 240.0,
  "y": 240.0,
  "rotation": 30.0,
  "period-millis": 2000,
  "color": "#ff0000ff",
  "effect": "explosion"
}
```

### Event Effect Object

###### Event Effect Object structure

> Event effects are such static effects, only without fields: <br>
> `x`, `y`, `period-millis`

| Name  | Description                             |
| ----- | --------------------------------------- |
| move  | effect appearing when the player moves  |
| join  | effect appearing when the player joins  |
| leave | effect appearing when the player leaves |

###### Example Event Effect Object

```json
{
  "rotation": 30.0,
  "color": "#4169e1ff",
  "effect": "fire"
}
```


src/main/java/inside/Config.java
package inside;

import mindustry.net.Host;

import java.util.*;

public class Config{

    /** Offline pattern, used when servers are offline. */
    public String offlinePattern = "\u26A0 [scarlet]Offline";

    /**
     * Online count pattern, used in servers status.
     * This is placeholder based patter, {@link LightweightHub#formatter}.
     * <p>
     * Syntax: <b>%fieldName%</b> - where fieldName is one of
     * a fields is {@link Host} class.
     */
    public String onlinePattern = "\uE837 [accent]Online: %players%\n\uE827 Map: %mapname%";

    /** If {@code true}, player connection logging will be enabled. */
    public boolean logConnects = false;

    /** List with {@link EffectData} that are triggered when the event is called. */
    public Map<String, EffectData> eventEffects = new HashMap<>();

    {
        eventEffects.put("move", new EffectData(0, 0, 30, 0, "#4169e1ff", "fire"));
        eventEffects.put("leave", new EffectData(0, 0, 30, 0, "#4169e1ff", "plasticExplosion"));
        eventEffects.put("join", new EffectData(0, 0, 30, 0, "#cbd97f", "instBomb"));
    }

    /** List with static {@link EffectData}. */
    public List<EffectData> effects = Arrays.asList(
            new EffectData(240, 240, 30, 2000, "#ff0000ff", "explosion"),
            new EffectData(160, 160, 30, 3000, "#ff0000ff", "explosion")
    );

    /** List with {@link HostData}. */
    public List<HostData> servers = Arrays.asList(
            new HostData(1000, 7, 22, 37, 200, 320, 200, 288),
            new HostData(2000, 7, 7, 22, 80, 200, 80, 168),
            new HostData(3000, 7, 22, 7, 200, 80, 200, 48),
            new HostData(4000, 7, 37, 23, 320, 200, 320, 168)
    );
}
